### ISSUE TITLE
EWYK concurrent produce can fail SSL connections

### ISSUE THREAD

**Issue by romrich** [eclipse/jetty.project#2468]
I'm using jetty server in the version 9.4.9.v20180320.
For some reasons after load tests some HTTPS connections can't be established:
`C:\>openssl s_client -connect localhost:9443
WARNING: can't open config file: /etc/ssl/openssl.cnf
CONNECTED(00000228)
`
As you can see I've got a TCP connection, but without SSL handshake.

My system environment: Windows 10 x64, JRE v1.8_162s

[threaddump_1.txt](https://github.com/eclipse/jetty.project/files/1928452/threaddump_1.txt)
[threaddump_2.txt](https://github.com/eclipse/jetty.project/files/1928453/threaddump_2.txt)

**Comment by joakime** [comment]
Can you explain how you have configured and started jetty in these tests?

**Comment by romrich** [comment]
`
 
    @Inject
    public SoapServer(Listener listener) throws Exception {
        _listener = listener;
        prepareServer();
    }

    private void prepareServer() throws Exception {
        System.setProperty("org.mortbay.http.Version.paranoid", "true");
        // _server.setTrace(false);
        QueuedThreadPool tp = new QueuedThreadPool();
        tp.setDaemon(true);
        tp.setName("MessagePort");
        // Create the server
        _server = new Server(tp);
        _server.setStopTimeout(1000);

        // Create a servlet container
        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        ErrorHandler paranoidErrorHandler = new ParanoidErrorHandler();
        context.setErrorHandler(paranoidErrorHandler);

        // Map a servlet onto the container
        String[] listeners = getConfig().getAllowedListenerTypes();
        ServletHolder holder = new ServletHolder(_listener);
        for (int i = 0; i < listeners.length; i++) {
            context.addServlet(holder, "/path/" + listeners[i]);
        }

        // Serve static content from the context
        ResourceHandler resource = new ResourceHandler();
        resource.setDirectoriesListed(false);
        resource.setResourceBase("webroot");
        HandlerList handlers = new HandlerList();
        handlers.setHandlers(new Handler[] { context, resource });
        StatisticsHandler stats = new StatisticsHandler();
        stats.setHandler(handlers);
        _server.setHandler(stats);
        MBeanContainer mbeanContainer = new MBeanContainer(ManagementFactory.getPlatformMBeanServer());
        _server.addBean(mbeanContainer);

        // Start the http server
        _server.start();
    }

    public void startHttpsPort() throws Exception {
        _log.debug("starting HTTPS server");

        final SecureListener instance = SecureListener.getInstance();
        SslContextFactory factory = new SslContextFactory(instance.getConfig().getServerKeystoreLocation());
        String[] allCiphers = SSLContext.getDefault().createSSLEngine().getSupportedCipherSuites();
        factory.setKeyManagerPassword(instance.getConfig().getPrivateKeyPassword());
        factory.setKeyStorePassword(SecureListener.DEFAULT_KEYSTORE_PASSWORD);
        factory.setTrustStore(buildTrustCAKeyStore(instance.getTrustKeystore()));
        factory.setKeyManagerFactoryAlgorithm(instance.getConfig().getKeystoreAlgorithm());
        factory.setTrustManagerFactoryAlgorithm(instance.getConfig().getKeystoreAlgorithm());
        factory.setRenegotiationAllowed(false);
        factory.addExcludeProtocols("SSLv2Hello", "SSLv3");

        // this is better for error handling
        factory.setNeedClientAuth(false);
        factory.setWantClientAuth(instance.getConfig().isSenderCertificateCheck());

        HttpConfiguration httpConfig = new HttpConfiguration();
        httpConfig.setSendServerVersion(false);
        HttpConnectionFactory httpFactory = new HttpConnectionFactory(httpConfig);

        _httpsSocket = new ServerConnector(_server, factory, httpFactory);
        _httpsSocket.setReuseAddress(true);
        _httpsSocket.setSoLingerTime(1000);

        String host = instance.getConfig().getSecureListenerHttpsAddress();
        int port = instance.getConfig().getSecureListenerHttpsPort();
        _httpsSocket.setName(host + ":" + port);

        _httpsSocket.setPort(port);
        if (!"*".equals(host)) {
            _httpsSocket.setHost(host);
        }
        _httpsSocket.setIdleTimeout(getConfig().getSecureListenerIdleConnectionTimeout());
        _server.addConnector(_httpsSocket);

        try {
            Enumeration<String> aliases = instance.getTrustKeystore().getRawKeystore().aliases();
            while (aliases.hasMoreElements()) {
                String alias = aliases.nextElement();
                if (instance.getTrustKeystore().getRawKeystore().isCertificateEntry(alias)) {
                    java.security.cert.X509Certificate cert = (java.security.cert.X509Certificate) instance.getTrustKeystore()
                            .getRawKeystore().getCertificate(alias);
                    if (!cert.getIssuerX500Principal().equals(cert.getSubjectX500Principal())) {
                        _log.debug("trusted certificate: " + cert.getSubjectDN().getName());
                    }
                }
            }
        } catch (KeyStoreException e) {
            _log.error("error during HTTPS server startup: " + e);
        }

        _httpsSocket.start();

        StatusSingleton.getInstance().setHttps(getAddressAndPort(_httpsSocket));
        StatusSingleton.getInstance().setHttpsClientCertCheck(instance.getConfig().isSenderCertificateCheck());
        _log.info("HTTPS server started on " + getAddressAndPort(_httpsSocket));
    }

`

**Comment by sbordet** [comment]
I looked at the stack traces and they show a condition that should not happen.

Namely, there are 2 threads in the select loop: one is inside the select operation, and the other is trying to register a new connection.
The former locks the key `Set` and keeps it locked while selecting; the latter is trying to lock the key `Set` but cannot, which makes this selector hung and that's why `openssl` is stuck.

Can you enable DEBUG logging and reproduce the issue ?
How long does it take to reproduce (i.e. DEBUG logs will be huge) ?
What client do you use to perform the load tests ?

FYI, NIO on Windows is not something we recommend; we have fought (in the past and recently) Windows NIO bugs that are hard to reproduce and to workaround.

**Comment by sbordet** [comment]
We are seeing this also on Linux in our CI systems.

**Comment by olamy** [comment]
https://gist.github.com/olamy/54203c6db053d7cb05c8732d1b47636d

**Comment by sbordet** [comment]
@romrich can you please tell whether in your logs there are lines logged at WARN level ?

**Comment by sbordet** [comment]
@romrich we were able to find a code path that reproduced the issue and fix it. 
I'm guessing you were hitting the same code path.
Can you please verify that commit 0a4d62c works for you ?

**Comment by romrich** [comment]
Unfortunately it looks not better.
I could see the follow WARNING:
`16:55:19.248 [MessagePort-239] WARN  HttpChannel - /pontonxp/SoapListener
java.lang.IllegalStateException: Partial message pending 80000 for BLOCKING
	at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.lockMsg(WebSocketRemoteEndpoint.java:128)
	at org.eclipse.jetty.websocket.common.WebSocketRemoteEndpoint.sendString(WebSocketRemoteEndpoint.java:386)
	at org.eclipse.jetty.websocket.jsr356.JsrBasicRemote.sendText(JsrBasicRemote.java:108)
	at de.ponton.securelistener.B.B.B.B(Unknown Source)
	at de.ponton.securelistener.B.B.B.A(Unknown Source)
	at de.ponton.securelistener.D.A.B.doPost(Unknown Source)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:865)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:169)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:531)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
	at org.eclipse.jetty.server.HttpChannel.run(HttpChannel.java:293)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:755)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:673)
	at java.lang.Thread.run(Thread.java:748)`

Our use case is the following. The jetty server receives data, which we forward through a opened websocket connection, initialized from a second application, to the second applications.
It seems, that we have a bug in our forward part, but the jetty server has to work further.
Attached you can find my whole log file and a thread dump.
[listener-2018-04-26[6].log.zip](https://github.com/eclipse/jetty.project/files/1954601/listener-2018-04-26.6.log.zip)
[threaddump_20180427.txt](https://github.com/eclipse/jetty.project/files/1954602/threaddump_20180427.txt)

PS: used jetty state: commit 97ec0b078de87fef0de523ac6cded7c1ca5c2ec3

**Comment by sbordet** [comment]
@romrich this is a different issue and related to WebSocket. Can you please open another issue ?

**Comment by romrich** [comment]
@sbordet no it isn't!
forget the websocket connections. The point is, that when we throw an exception java.lang.IllegalStateException in the doPost(), the jetty server doesn't work anymore.

In the thread dump you can see, that there are no running MessagePort threads anymore. This is the jetty server bug.

**Comment by romrich** [comment]
@sbordet normally if I start my application I have four MessagePort threads in RUNNUNG state and one MessagePort-x-acceptor thread in RUNNIING state two.
But I hadn't these four running threads anymore:
`Thread Name	Thread State	Blocked Count	Blocked Time	Waited Count	Waited Time	Total CPU Usage(%)	Deadlocked	Lock Name	Lock Owner ID	Lock Owner Name	Thread Id	Native	Suspended	Allocated Bytes(bytes)
MessagePort-231	TIMED_WAITING	30		1.978		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			231	0	0	-256.236
MessagePort-163	TIMED_WAITING	37		1.932		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			163	0	0	-256.236
MessagePort-138	TIMED_WAITING	38		1.993		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			138	0	0	-256.236
MessagePort-129	TIMED_WAITING	31		1.976		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			129	0	0	-256.236
MessagePort-110	TIMED_WAITING	40		1.979		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			110	0	0	-256.236
MessagePort-98	TIMED_WAITING	40		1.966		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			98	0	0	-256.236
MessagePort-82	TIMED_WAITING	37		1.985		-45.623.600	Not Enabled	java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@33f948f6			82	0	0	-256.236
MessagePort-33-acceptor-0@3101663c-*:9443@19e119eb{SSL,[ssl, http/1.1]}{0.0.0.0:9443}	RUNNABLE	0		1		-45.623.600	Not Enabled				33	1	0	-256.236
`
When I have this state it's not possible to open SSL connection to the jetty server anymore. I just get only a tcp connection:

c:\>openssl s_client -connect localhost:9443
WARNING: can't open config file: /etc/ssl/openssl.cnf
CONNECTED(00000224)
write:errno=10053
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 307 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1524827500
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---

c:\>

**Comment by romrich** [comment]
@sbordet I test with the jMeter

**Comment by joakime** [comment]
_(This isn't related to your main Issue, just the "Partial message pending 80000 for BLOCKING")_

You might want to update your `doPost()` to JSR `sendText()` path.
You have multiple outside threads attempting to use blocking send (only 1 thread can block, the one that is sending)

Advice:

* There should be a queue in front of your `sendText()`
* The `doPost()` offers/adds to the queue.
* The implementation on the JSR side should poll/pull from the queue and send only when the prior send is complete.
* The queue should have upper limits to know if you are overloading the websocket endpoint or not. (could be a slow remote, a bad remote, or even a flood of message from doPost that the websocket connection cannot handle.
* The queue is also a good place to add priorities on the messages. (important ones bubble up, less important ones can be dropped if the queue is too large, etc)

Note that there is also an Async Remote that can provide information about the send result.
Check javax.websocket.RemoteEndpoint.Async to utilize the `sendText(String, SendHandler)` method to gain knowledge about send result.
You would use the javax.websocket.SendResult provided in the javax.websocket.SendHandler.onResult(SendResult) to know if the send finished or not.

**Comment by joakime** [comment]
_(This also isn't related to your main Issue, just the "WARNING: can't open config file: /etc/ssl/openssl.cnf")_

The warning/error you see  ...

```
c:\>openssl s_client -connect localhost:9443
WARNING: can't open config file: /etc/ssl/openssl.cnf
```

means you haven't properly finished the installation and configuration of openssl on Windows.
This configuration is important for the kinds of testing you are doing.

Per the openssl for windows installation documentation you still have to manually create/setup the `OPENSSL_CONF` system environment variable.
It should point to the installed copy of your openssl.cnf file.

Some example `OPENSSL_CONF` values:

* `C:\Program Files (x86)\GnuWin32\share\openssl.cnf` - if cygwin or gnu version installed
* `C:\OpenSSL-Win32\bin\openssl.cfg` - if native 32-bit installed
* `C:\OpenSSL-Win64\bin\openssl.cfg` - if native 64-bit installed
* `C:\Program Files (x86)\Git\ssl\openssl.cfg` - if using it from native git (32-bit)
* etc ..

**Comment by sbordet** [comment]
@romrich why do you have only 4 `MessagePort` threads ? Are you underconfiguring the thread pool ?

The last thread dump you reported shows 4 selectors, 1 acceptor, 8 idle threads and 2 reserved threads.
It seems to me an idle server waiting for requests.

Can you please configure the server with:
```java
queuedThreadPool.setDetailedDump(true);
```
and then take a Server dump as described here: https://www.eclipse.org/jetty/documentation/9.4.x/jetty-dump-tool.html

**Comment by romrich** [comment]
@sbordet normally there are more than 4 MessagePort threads, but 4 of them are in state RUNNING (shown in JMC). After load tests there are no threads in the state RUNNING and the jetty server doesn't work correctly anymore.
I try to make a thread dump with the detailedDump flag.

**Comment by romrich** [comment]
@sbordet sorry for my mistake. The state, which I mentioned, is RUNNABLE and not RUNNING.

**Comment by sbordet** [comment]
@romrich please provide a `Server` dump and also detail what is your configuration, especially for the thread pool.

**Comment by romrich** [comment]
@sbordet here are the requested dump files:
[server_dump_20180430.txt](https://github.com/eclipse/jetty.project/files/1959967/server_dump_20180430.txt)
[threaddump_20180430.txt](https://github.com/eclipse/jetty.project/files/1959968/threaddump_20180430.txt)

**Comment by romrich** [comment]
private void prepareServer() throws Exception {
        System.setProperty("org.mortbay.http.Version.paranoid", "true");
        // _server.setTrace(false);
        QueuedThreadPool tp = new QueuedThreadPool();
        tp.setDetailedDump(true);
        tp.setDaemon(true);
        tp.setName("MessagePort");
        // Create the server
        _server = new Server(tp);
        _server.setStopTimeout(1000);

        // Create a servlet container
        ServletContextHandler context = new ServletContextHandler();
        context.setContextPath("/");
        ErrorHandler paranoidErrorHandler = new ParanoidErrorHandler();
        context.setErrorHandler(paranoidErrorHandler);

        // Map a servlet onto the container
        String[] listeners = getConfig().getAllowedListenerTypes();
        ServletHolder holder = new ServletHolder(_listener);
        for (int i = 0; i < listeners.length; i++) {
            context.addServlet(holder, "/pontonxp/" + listeners[i]);
        }

        // Serve static content from the context
        ResourceHandler resource = new ResourceHandler();
        resource.setDirectoriesListed(false);
        resource.setResourceBase("webroot");
        HandlerList handlers = new HandlerList();
        handlers.setHandlers(new Handler[] { context, resource });
        StatisticsHandler stats = new StatisticsHandler();
        stats.setHandler(handlers);
        _server.setHandler(stats);
        MBeanContainer mbeanContainer = new MBeanContainer(ManagementFactory.getPlatformMBeanServer());
        _server.addBean(mbeanContainer);

        // Start the http server
        _server.start();
    }

**Comment by sbordet** [comment]
@romrich your latest code and dumps shows that there is no `ServerConnector`, so the server does not listen to any port.

**Comment by romrich** [comment]
a ServerConnector is started in the other method:

    public void startHttpsPort() throws Exception {
        _log.debug("starting HTTPS server");

        final SecureListener instance = SecureListener.getInstance();
        SslContextFactory factory = new SslContextFactory(instance.getConfig().getServerKeystoreLocation());
        String[] allCiphers = SSLContext.getDefault().createSSLEngine().getSupportedCipherSuites();
        factory.setKeyManagerPassword(instance.getConfig().getPrivateKeyPassword());
        factory.setKeyStorePassword(SecureListener.DEFAULT_KEYSTORE_PASSWORD);
        factory.setTrustStore(buildTrustCAKeyStore(instance.getTrustKeystore()));
        factory.setKeyManagerFactoryAlgorithm(instance.getConfig().getKeystoreAlgorithm());
        factory.setTrustManagerFactoryAlgorithm(instance.getConfig().getKeystoreAlgorithm());
        factory.setRenegotiationAllowed(false);
        factory.addExcludeProtocols("SSLv2Hello", "SSLv3");

        // this is better for error handling
        factory.setNeedClientAuth(false);
        factory.setWantClientAuth(instance.getConfig().isSenderCertificateCheck());

        HttpConfiguration httpConfig = new HttpConfiguration();
        httpConfig.setSendServerVersion(false);
        HttpConnectionFactory httpFactory = new HttpConnectionFactory(httpConfig);

        _httpsSocket = new ServerConnector(_server, factory, httpFactory);
        _httpsSocket.setReuseAddress(true);
        _httpsSocket.setSoLingerTime(1000);

        String host = instance.getConfig().getSecureListenerHttpsAddress();
        int port = instance.getConfig().getSecureListenerHttpsPort();
        _httpsSocket.setName(host + ":" + port);

        _httpsSocket.setPort(port);
        if (!"*".equals(host)) {
            _httpsSocket.setHost(host);
        }
        _httpsSocket.setIdleTimeout(getConfig().getSecureListenerIdleConnectionTimeout());
        _server.addConnector(_httpsSocket);

        try {
            Enumeration<String> aliases = instance.getTrustKeystore().getRawKeystore().aliases();
            while (aliases.hasMoreElements()) {
                String alias = aliases.nextElement();
                if (instance.getTrustKeystore().getRawKeystore().isCertificateEntry(alias)) {
                    java.security.cert.X509Certificate cert = (java.security.cert.X509Certificate) instance.getTrustKeystore()
                            .getRawKeystore().getCertificate(alias);
                    if (!cert.getIssuerX500Principal().equals(cert.getSubjectX500Principal())) {
                        _log.debug("trusted certificate: " + cert.getSubjectDN().getName());
                    }
                }
            }
        } catch (KeyStoreException e) {
            _log.error("error during HTTPS server startup: " + e);
        }

        _httpsSocket.start();

        StatusSingleton.getInstance().setHttps(getAddressAndPort(_httpsSocket));
        StatusSingleton.getInstance().setHttpsClientCertCheck(instance.getConfig().isSenderCertificateCheck());
        _log.info("HTTPS server started on " + getAddressAndPort(_httpsSocket));
    }

**Comment by romrich** [comment]
@sbordet just search for the port 9443:

"MessagePort-34-acceptor-0@2aac8b32-*:9443@7d64acaf{SSL,[ssl, http/1.1]}{0.0.0.0:9443}" #34 daemon prio=3 os_prio=-1 tid=0x0000000019c9f000 nid=0x6048 runnable [0x000000001c8df000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method)
	at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422)
	at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250)
	- locked <0x00000000f0739dc8> (a java.lang.Object)
	at org.eclipse.jetty.server.ServerConnector.accept(ServerConnector.java:378)
	at org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:639)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:755)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:673)
	at java.lang.Thread.run(Thread.java:748)

   Locked ownable synchronizers:
	- None

**Comment by sbordet** [comment]
@romrich the server dump shows that there is no `ServerConnector` bean attached to the server.

If you took the dump at the moment the server hung, then I should be able to see the `ServerConnector` as a bean, and I should be able to see `ManagedSelector` beans as well.
They are both missing from the dump.

Do you have `WARN` messages or exceptions in your logs ?

I don't understand how you can have the acceptor thread running, but not a single selector thread.

**Comment by romrich** [comment]
@sbordet if you have time for a remote session (per GoToMeeting) I can show you, how I start the jetty server and test it?

no WARN or execptions are logged.

**Comment by sbordet** [comment]
Nope. When are you taking the dump ? When the system is hung ?

**Comment by romrich** [comment]
@sbordet I'll take three thread/server dumps: after start, after some processed requests, when the jetty server hung

PS: the last dump was done, when the jetty server is hanged

**Comment by romrich** [comment]
@sbordet here are the dump files:
after the start:
[server_dump_20180430_1.txt](https://github.com/eclipse/jetty.project/files/1960338/server_dump_20180430_1.txt)
[serverconnector_dump_20180430_1.txt](https://github.com/eclipse/jetty.project/files/1960339/serverconnector_dump_20180430_1.txt)
[threaddump_20180430_1.txt](https://github.com/eclipse/jetty.project/files/1960340/threaddump_20180430_1.txt)

after some (>1000) received/processed messages:
[server_dump_20180430_2.txt](https://github.com/eclipse/jetty.project/files/1960345/server_dump_20180430_2.txt)
[serverconnector_dump_20180430_2.txt](https://github.com/eclipse/jetty.project/files/1960346/serverconnector_dump_20180430_2.txt)
[threaddump_20180430_2.txt](https://github.com/eclipse/jetty.project/files/1960347/threaddump_20180430_2.txt)

after the jetty server is hanged:
[server_dump_20180430_3.txt](https://github.com/eclipse/jetty.project/files/1960348/server_dump_20180430_3.txt)
[serverconnector_dump_20180430_3.txt](https://github.com/eclipse/jetty.project/files/1960349/serverconnector_dump_20180430_3.txt)
[threaddump_20180430_3.txt](https://github.com/eclipse/jetty.project/files/1960350/threaddump_20180430_3.txt)

**Comment by sbordet** [comment]
@romrich bizarre.

The `ServerConnector` dumps shows that when the system is hung, the NIO selector is gone:

```
 |   += org.eclipse.jetty.io.ManagedSelector@314c3726 id=0 keys=-1 selected=-1 updates=1 - STARTED
```

The fact that the `ManagedSelector` is started, but `keys=-1` means that the NIO selector threw an exception and was closed.

Gut feeling is that it's another Windows NIO bug, as the selector should not throw exceptions.
However, if it does we should handle the situation better.

You must have the exception logged in your server logs at WARN level. Make sure your logs are enabled for that level. Can you report the exception here ?

@gregw apparently an exception is thrown in `ManagedSelector.select()`, we catch them all, close the NIO selector and null out `_selector`. Perhaps we can allocate a new one rather than nulling it out ?

@romrich would you be able to test the latest Jetty code if we put in some a recovery attempt when the selector throws an exception ?

**Comment by sbordet** [comment]
@romrich the class that will log the warning when the NIO selector throws an exception is under the `org.eclipse.jetty.io` logger name/category.
Can you please verify that you are logging at least at WARN level for that category ?

**Comment by gregw** [comment]
@romrich The code that we are concerned with is https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java#L407-L415
The log warn on line 412 should report the issue.   If you can't see that log, can you modify the code to do a `x.printStackTrace()` or similar so we can see what the underlying problem is?

@sbordet recreating the selector may well be a work around.... or it might be an infinite loop.  We need to see what the exception is first

**Comment by romrich** [comment]
@gregw and @sbordet Here is the logged exception:
```
10:58:05.621 [MessagePort-67] WARN  ManagedSelector - 
java.io.IOException: Ein nicht blockierender Socketvorgang konnte nicht sofort ausgeführt werden
	at sun.nio.ch.SocketDispatcher.close0(Native Method)
	at sun.nio.ch.SocketDispatcher.close(SocketDispatcher.java:63)
	at sun.nio.ch.SocketChannelImpl.kill(SocketChannelImpl.java:879)
	at sun.nio.ch.WindowsSelectorImpl.implDereg(WindowsSelectorImpl.java:588)
	at sun.nio.ch.SelectorImpl.processDeregisterQueue(SelectorImpl.java:149)
	at sun.nio.ch.WindowsSelectorImpl.doSelect(WindowsSelectorImpl.java:142)
	at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
	at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:101)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.select(ManagedSelector.java:396)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:357)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:181)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:755)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:673)
	at java.lang.Thread.run(Thread.java:748)
```
please see
[jetty-2018-05-02[25].log.zip](https://github.com/eclipse/jetty.project/files/1966812/jetty-2018-05-02.25.log.zip)
[jetty.zip](https://github.com/eclipse/jetty.project/files/1966813/jetty.zip)

**Comment by sbordet** [comment]
The exception message says: "A non-blocking socket operation could not be performed immediately"

**Comment by joakime** [comment]
That specific text "Ein nicht blockierender Socketvorgang konnte nicht sofort ausgeführt werden" ...

Comes from Microsoft's own Windows Socket code for Error 10035 is described as:

WSAEWOULDBLOCK
10035	A non-blocking socket operation could not be completed immediately

**Comment by romrich** [comment]
@sbordet and @joakime Just for your information. With jetty version 9.3.23.v20180228 I don't have these kind of problems

**Comment by mhaeusler** [comment]
Any news on this ticket ?
I am facing the same problem :(

**Comment by sbordet** [comment]
@romrich @mhaeusler I dug a bit more and I found that `SocketDispatcher.close0()` eventually calls the Windows API `closesocket()`, which is described here: https://msdn.microsoft.com/en-us/library/windows/desktop/ms737582(v=vs.85).aspx

There, it reports that `WSAEWOULDBLOCK` is returned if the socket is configured with `SoLinger`.

Can you please remove your setting for the linger time, and try again?

**Comment by sbordet** [comment]
Confirmed that linger should not be set for NIO sockets:
http://mail.openjdk.java.net/pipermail/nio-dev/2018-June/005195.html

We should deprecate and possibly remove the settings for linger in Jetty, as we now only use NIO sockets.
@gregw your opinion?

**Comment by gregw** [comment]
@sbordet can  we close this issue now?

**Comment by sbordet** [comment]
@gregw I think so, although would be great if @romrich and @mhaeusler confirm that their problem was indeed caused by setting `soLinger`.
They can eventually open another issue.

**Comment by mhaeusler** [comment]
yes, I can confirm that the problem does not occur when soLinger is not set.
however it is still strange that the problem did not show up when using jetty 9.3 or earlier.

**Comment by gregw** [comment]
@mhaeusler it is indeed strange, but we have changed aspects of the code that calls the selector, specifically which threads do the calls and in what order.   We've also added extra calls to selectNow to work around other windows selector bugs.   So it is likely that something in these changes calling sequences has triggered the windows bug.  I too would like to know exactly why, but without debugging the internals of windows I think we'll have to be satisfied with knowing how to avoid it.
